### ISSUE TITLE
[spark] Make xgboost.spark support spark connect ML

### ISSUE THREAD

**Issue by WeichenXu123** [dmlc/xgboost#9780]
Since spark 3.5, a new pyspark module is added: `pyspark.ml.connect`, it supports a few ML algorithms that runs on spark connect mode. This is design doc:
https://www.google.com/url?q=https://docs.google.com/document/d/1LHzwCjm2SluHkta_08cM3jxFSgfF-niaCZbtIThG-H8/edit&sa=D&source=calendar&ust=1700005806011038&usg=AOvVaw2VEdVyMYg40yDLpElhcRAu

We should make estimators defined in `xgboost.spark` to support spark connect mode, to achieve the goal, we need:

- Make these estimator class inherits `pyspark.ml.connect.Estimator` if it runs on spark connect mode
- All implementation code should only calls spark connect API (i.e. spark Dataframe API).

**Comment by WeichenXu123** [comment]
CC @wbo4958

**Comment by wbo4958** [comment]
Yes, that is a good suggestion. However, I have a concern that the spark dataframe API hasn't supported stage-level scheduling yet. in that case, do we need to force only 1 task running on the executor?

**Comment by WeichenXu123** [comment]
> Yes, that is a good suggestion. However, I have a concern that the spark dataframe API hasn't supported stage-level scheduling yet. in that case, do we need to force only 1 task running on the executor?

It is supported, see:

https://spark.apache.org/docs/latest/api/python/reference/pyspark.sql/api/pyspark.sql.DataFrame.mapInPandas.html?highlight=mapinpandas

and

https://spark.apache.org/docs/latest/api/python/reference/pyspark.sql/api/pyspark.sql.DataFrame.mapInArrow.html?highlight=mapinarrow#pyspark.sql.DataFrame.mapInArrow

New `barrier` argument is added for them.
@wbo4958

**Comment by wbo4958** [comment]
Hi @WeichenXu123, I mean the stage-level scheduling not the barrier execution. I guess we can support DataFrame stage-level scheduling in Spark for specific APIs like mapInPandas and mapInArrow using the same way as the barrier supporting. That will be cool.

**Comment by wbo4958** [comment]
BTW, I'd like to take this task to make xgboost.spark support spark connect ML.

**Comment by WeichenXu123** [comment]
> Hi @WeichenXu123, I mean the stage-level scheduling not the barrier execution. I guess we can support DataFrame stage-level scheduling in Spark for specific APIs like mapInPandas and mapInArrow using the same way as the barrier supporting. That will be cool.

oh, sorry for my misread, yes we haven't support stage-level scheduling in spark connect api, this is a todo task

**Comment by WeichenXu123** [comment]
We can add stage-level scheduling params in `mapInPandas` API, similar to `barrier` param, CC @zhengruifeng @Ngone51 WDYT ?

**Comment by zhengruifeng** [comment]
> We can add stage-level scheduling params in `mapInPandas` API, similar to `barrier` param, CC @zhengruifeng @Ngone51 WDYT ?

I think this way is feasible.

**Comment by wbo4958** [comment]
Cool, let me have the PR supporting stage-level scheduling for Dataframe API for spark @WeichenXu123 @zhengruifeng @Ngone51

**Comment by trivialfis** [comment]
Hi, may I ask what's the current status of this?

**Comment by medb** [comment]
Is this still relevant after Spark 4.0 release that added SparkML support to Spark Connect?
